lib/spack/spack/package.py

```diff
@@ -2195,6 +2195,85 @@ def possible_dependencies(*pkg_or_spec, **kwargs):
     return visited
 
 
+#: Registers which are the detectable packages, by repo and package name
+detectable_packages = collections.defaultdict(list)
+
+
+def detectable(decorated_cls):
+    """Mark a package class as detectable and add a default implementation
+    for detection logic.
+
+    The decorator registers the class as detectable and attaches to it a
+    default implementation of ``determine_spec_details``.
+
+    Args:
+        decorated_cls: class to be decorated
+
+    Returns:
+        Decorated class
+    """
+    @classmethod
+    def determine_spec_details(cls, prefix, exes_in_prefix):
+        """Allow ``spack external find ...`` to locate installations.
+
+        Args:
+            prefix (str): the directory containing the executables
+            exes_in_prefix (set): the executables that match the regex
+
+        Returns:
+            The list of detected specs for this package
+        """
+        exes_by_version = collections.defaultdict(list)
+        # The default filter function is the identity function for the
+        # list of executables
+        filter_fn = getattr(cls, 'filter_detected_exes', lambda x, exes: exes)
+        exes_in_prefix = filter_fn(prefix, exes_in_prefix)
+        for exe in exes_in_prefix:
+            try:
+                version_str = cls.determine_version(exe)
+                if version_str:
+                    exes_by_version[version_str].append(exe)
+            except Exception as e:
+                msg = ('An error occurred when trying to detect the version '
+                       'of "{0}" [{1}]')
+                tty.debug(msg.format(exe, str(e)))
+
+        specs = []
+        for version_str, exes in exes_by_version.items():
```

Comment:
What about exes where we weren't able to determine the version? Those need to be handled too.

Reply:
Determining a version is required in this default implementation. An executable for which we weren't able to determine a version won't pass to the `determine_variants` stage. The rationale is that we don't want to register something with no version as an external because, once we have proper constraint resolution, it means that it will satisfy *every* version request - which wouldn't obviously be true in reality.

Reply:
I disagree with this approach. In #16691 I found that the system libtool on macOS doesn’t have a way to query the version. In fact, a lot of macOS system tools have their own versioning scheme that doesn’t match the packages in Spack. I think we should be more robust and allow external packages without versions. We should also allow them to determine variants even if no version was detected.

Reply:
> I think we should be more robust and allow external packages without versions

We already allow that, but it won't be the default. If you need something like that you need to override `determine_spec_details` instead of relying on the default implementation and override simpler callbacks. 

My worry is that it's far too common that package A at version X needs package B at version greater than Y to build. Think if you have `cmake` at version 2.8 but can't detect the version. How many packages would break if you just register `cmake` because they need `cmake@3:`?

Reply:
I understand that it's allowed, but I would like it to be the default `determine_spec_details` implementation.

In my mind, the point of `spack external find` is to simplify the process of manually adding external packages. Users still have to edit `packages.yaml` by hand to add `buildable: False`. While they're at it, they can check versions and variants and update them if they need to.

@tgamblin Will the new concretizer be smart enough to understand that if a package requires `foo@3:` and we have `foo` installed, that it may or may not satisfy this constraint? This is similar to the package rebuild question of whether `foo` satisfies `foo+bar` or `foo~bar`, and the answer should be neither. In @alalazo's example, I think it makes sense for the concretizer to use an external `cmake` if no version requirement is specified, and for Spack to build its own `cmake` if it requires `cmake@3:` and we don't know the version.

Reply:
> Users still have to edit packages.yaml by hand to add buildable: False. While they're at it, they can check versions and variants and update them if they need to.

Users can ask to find something and mark it not buildable from command line:
```console
$ spack external find --not-buildable cmake
```

Reply:
> This is similar to the package rebuild question of whether foo satisfies foo+bar or foo~bar, and the answer should be neither.

I think it should be both, if you consider it a constraint :thinking:

Reply:
If a package depends on `foo`, then it should be satisfied by the following external packages: `foo`, `foo@2:`, `foo@3+bar`

If a package depends on `foo@2:+bar` or `foo@2:~bar`, it should not be satisfied by any of the above external packages. Spack shouldn't guess versions or variants if it doesn't know them.

Reply:
> If a package depends on foo@2:+bar or foo@2:~bar, it should not be satisfied by any of the above external packages. Spack shouldn't guess versions or variants if it doesn't know them.

This is a complete change on the current semantics:
```console
$ spack external find cmake
$ cat ~/.spack/linux/packages.yaml
packages:
  cmake:
    externals:
    - spec: cmake@3.17.2
      prefix: /usr
    buildable: false
$ spack install cmake+ncurses
==> cmake@3.17.2 : externally installed in /usr
==> cmake@3.17.2 : generating module file
==> cmake@3.17.2 : registering into DB
$ spack install cmake~ncurses
==> cmake@3.17.2 : externally installed in /usr
==> cmake@3.17.2 : already registered in DB
```

Reply:
I understand that, I don't like the current semantics.

Reply:
@adamjstewart Ah, ok. I'm totally fine discussing how to change this behavior, maybe though I'd save it for another PR. Do you agree that if we keep the current behavior, requiring to find a version is a sensible default?

Reply:
Even with the current concretizer behavior (which will hopefully be changed once the new concretizer comes... soon?), I would prefer to not require a version. But this PR is clearly an improvement over the current state of things, so if I'm the only one who thinks this then I would rather see this merged than delayed for me.